### ISSUE TITLE
Way to directly create decimals with scale without dividing?

### ISSUE THREAD

**Issue by tylerhawkes** [neogenie/fastnum#3]
For some algo backtests I store data in postgres as i32 because it's much more compact and I know where the decimal point is for every column in the table. I'm trying to create a D128 from an i32 that I've loaded and there is no easy way to set the scale without forcing division. The `new` function is private and takes a `ControlBlock`.

The BigDecimal crate has the ability to directly create this using a function. Would you be against adding something that allows rescaling without any other side effects? If it exists, I can't find it in the code or docs.

I'm also running into issues converting from UnsignedDecimal to Decimal and it feels like there should be a simple conversion using the `From` trait there since I think they can both represent the same range of values.

**Comment by tylerhawkes** [comment]
I found the `from_parts` function, but it is behind the `dev` feature.

**Comment by neogenie** [comment]
@tylerhawkes 
Thank You for the feedback!

I can add the necessary constructor that accepts an integer part and scale.

Right now you can use two approaches:

1. `.from_parts()` is designed to create a decimal from components for external crates during integration (For example, when `diesel` integration will be moved from this crate to `diesel`).
2. `D128::quantum(exp, Context::default()) * digits`- this can be convenient in some cases and relatively fast, because only includes integer multiplication by `1`.

**Comment by neogenie** [comment]
> I'm also running into issues converting from UnsignedDecimal to Decimal and it feels like there should be a simple conversion using the From trait there since I think they can both represent the same range of values.

Sure! I'll add `From` and `TryFrom` implementations.

**Comment by tylerhawkes** [comment]
> 1. `.from_parts()` is designed to create a decimal from components for external crates during integration (For example, when `diesel` integration will be moved from this crate to `diesel`).

I think this is useful for more than just a few crates and being able to create these from scratch can help when converting custom data types. I don't think it needs to be behind a feature.



> 2\. `D128::quantum(exp, Context::default()) * digits`- this can be convenient in some cases and relatively fast, because only includes integer multiplication by `1`.

Thanks for pointing that out. the rust_decimal and bigdecimal crates have a way to rescale, where it is guaranteed to only change the scale. I think that having a `with_scale` function that takes by value is nice for chaining.

I got further along integrating and realized that there aren't any exponentiation functions. Do you have plans to add them? It turns out I need to raise to a power to find a geometric mean and find square roots (which is just a special case of raising to a half power).

**Comment by neogenie** [comment]
1. [v0.1.9](https://github.com/neogenie/fastnum/releases/tag/v0.1.9) with public [`::from_parts()`](https://docs.rs/fastnum/latest/fastnum/decimal/struct.Decimal.html#method.from_parts) is published.
2. https://docs.rs/fastnum/latest/fastnum/#rescale

> I got further along integrating and realized that there aren't any exponentiation functions. Do you have plans to add them? It turns out I need to raise to a power to find a geometric mean and find square roots (which is just a special case of raising to a half power).

The main theme of the library is performance. 
Unfortunately, exponentiation and calculation of various logarithms in reasonable time is not a trivial task. 
The current status is WIP. So far, only the function of raising to an integer power has been implemented: https://docs.rs/fastnum/latest/fastnum/decimal/struct.Decimal.html#method.powi

**Comment by tylerhawkes** [comment]
Yeah, I'm finding major performance issues having a stab at the rust_decimal crate. It's like 200x slower than using f64's. I honestly would rather have something that works and isn't super optimized and maybe has a warning about it being slower than expected sooner and then the implementation can be iterated on.



> 2\. https://docs.rs/fastnum/latest/fastnum/#rescale

I'm aware of that, but if I use `Decimal::ONE` and then rescale it to 5 that will give me 1.00000, what I'm looking for is something that will just change it to 0.00001 without changing the coefficient and just affecting the scale and guaranteeing (as much as possible) that it will only ever affect the scale.

**Comment by neogenie** [comment]
> I'm aware of that, but if I use Decimal::ONE and then rescale it to 5 that will give me 1.00000, what I'm looking for is something that will just change it to 0.00001 without changing the coefficient and just affecting the scale and guaranteeing (as much as possible) that it will only ever affect the scale.

Can you give an example of using this approach and describe the task in more detail?

**Comment by tylerhawkes** [comment]
> Can you give an example of using this approach and describe the task in more detail?

After thinking about it, this is only useful at the boundary when integrating with other types, which the `from_parts()` function does. So no real use case for it right now.
